sys-utils/setpriv-landlock.c

```diff
@@ -0,0 +1,211 @@
+	{ LANDLOCK_ACCESS_FS_REFER,       "refer"       },
+#endif
+#ifdef LANDLOCK_ACCESS_FS_TRUNCATE
+	{ LANDLOCK_ACCESS_FS_TRUNCATE,    "truncate"    },
+#endif
+};
+
+static long landlock_access_to_mask(const char *str, size_t len)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(landlock_access_fs); i++)
+		if (strncmp(landlock_access_fs[i].type, str, len) == 0)
+			return landlock_access_fs[i].value;
+	return -1;
+}
+
+static uint64_t parse_landlock_fs_access(const char *list)
+{
+	unsigned long r = 0;
+	size_t i;
+
+	/* without argument, match all */
+	if (list[0] == '\0') {
+		for (i = 0; i < ARRAY_SIZE(landlock_access_fs); i++)
+			r |= landlock_access_fs[i].value;
+	} else {
+		if (string_to_bitmask(list, &r, landlock_access_to_mask))
+			errx(EXIT_FAILURE,
+			     _("could not parse landlock fs access: %s"), list);
+	}
+
+	return r;
+}
+
+void parse_landlock_access(struct setpriv_landlock_opts *opts, const char *str)
+{
+	const char *type;
+	size_t i;
+
+	if (strcmp(str, "fs") == 0) {
+		for (i = 0; i < ARRAY_SIZE(landlock_access_fs); i++)
+			opts->access_fs |= landlock_access_fs[i].value;
+		return;
+	}
+
+	type = startswith(str, "fs:");
+	if (type)
+		opts->access_fs |= parse_landlock_fs_access(type);
+}
+
+void parse_landlock_rule(struct setpriv_landlock_opts *opts, const char *str)
+{
+	struct landlock_rule_entry *rule = xmalloc(sizeof(*rule));
+	const char *accesses, *path;
+	char *accesses_part;
+	int parent_fd;
+
+	accesses = startswith(str, "path-beneath:");
+	if (!accesses)
+		errx(EXIT_FAILURE, _("invalid landlock rule: %s"), str);
+	path = strchr(accesses, ':');
+	if (!path)
+		errx(EXIT_FAILURE, _("invalid landlock rule: %s"), str);
+	rule->rule_type = LANDLOCK_RULE_PATH_BENEATH;
+
+	accesses_part = xstrndup(accesses, path - accesses);
+	rule->path_beneath_attr.allowed_access = parse_landlock_fs_access(accesses_part);
+	free(accesses_part);
+
+	path++;
+
+	parent_fd = open(path, O_RDONLY | O_PATH | O_CLOEXEC);
+	if (parent_fd == -1)
+		err(EXIT_FAILURE, _("could not open file for landlock: %s"), path);
+
+	rule->path_beneath_attr.parent_fd = parent_fd;
+
+	list_add(&rule->head, &opts->rules);
+}
+
+void init_landlock_opts(struct setpriv_landlock_opts *opts)
+{
+	INIT_LIST_HEAD(&opts->rules);
+}
+
+void do_landlock(const struct setpriv_landlock_opts *opts)
+{
+	struct landlock_rule_entry *rule;
+	struct list_head *entry;
+	int fd, ret;
+
+	const struct landlock_ruleset_attr ruleset_attr = {
+		.handled_access_fs = opts->access_fs,
+	};
+
+	fd = landlock_create_ruleset(&ruleset_attr, sizeof(ruleset_attr), 0);
+	if (fd == -1)
+		err(SETPRIV_EXIT_PRIVERR, _("landlock_create_ruleset failed"));
+
+	list_for_each(entry, &opts->rules) {
+		rule = list_entry(entry, struct landlock_rule_entry, head);
+
+		assert(rule->rule_type == LANDLOCK_RULE_PATH_BENEATH);
+
+		ret = landlock_add_rule(fd, rule->rule_type, &rule->path_beneath_attr, 0);
+		if (ret == -1)
+			err(SETPRIV_EXIT_PRIVERR, _("adding landlock rule failed"));
```

Comment:
It would be nice to be able to specify a best-effort approach to still enable partial restriction even if the (old) running kernel doesn't support all util-linux's known access rights. For instance, with the current approach, not specifying `--landlock-access` might work on the user's system with one version of util-linux but not with a newer version. Likewise, users using an old kernel that only supports a subset of `landlock_access_fs` would not be able to use Landlock with setpriv unless they specify `--landlock-access`.

Reply:
I guess a new CLI flag would fit nicely here.